WindowsUsbDevice#getControllerDeviceIdList is not thread safe

I'm marking this issue for [first-timers-only](https://www.firsttimersonly.com/).  That means that I will only accept a PR for this one from someone who's never contributed to open source before. This one is easy (but don't make that statement make you feel bad if you have a hard time with it, there's more to contributing to open source than changing lines of code, especially if it's your first time). I'll hold your hand through this if you need me to. :-) 

As noted in [this report](https://scrutinizer-ci.com/g/oshi/oshi/issues/master/files/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java?selectedLabels%5B0%5D=37&orderField=path&order=asc&honorSelectedPaths=0&issueId=40218079#inspectioncomment-136359521), the lazy initialization of a list in `WindowsUSBDevice` is not thread safe.  If two threads call `getUsbDevices()` simultaneously they could both pass the `null` check and then both attempt to modify the same variable.

Create a thread-safe lazy initialization using OSHI's `Memoizer` class.

Here are the steps to get a PR merged here.
- Read the [CONTRIBUTING](https://github.com/oshi/oshi/blob/master/CONTRIBUTING.md) file for reference.  I'll walk you through anything you don't understand.
- Fork the project to your own GitHub account, make a local clone, and create a branch for your submission.
- In the [`WindowsUsbDevice`](https://github.com/oshi/oshi/blob/master/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java) class:
  - Modify the [`getControllerDeviceIdList()`](https://github.com/oshi/oshi/blob/31858912752ada0e0ceb43b43196ed1409601b96/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java#L309) method to take no arguments.  Inside the method, get the `wmiQueryHandler` by [copying from line 136](https://github.com/oshi/oshi/blob/31858912752ada0e0ceb43b43196ed1409601b96/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java#L136).  
  - Also remove the `if (controllerDeviceIds == null)` conditional 
  - Redefine that `controllerDeviceIds` variable with a differently named local variable in the method. 
  - Replace the static class variable with a memoized supplier: `private static Supplier<List<String>> controllerDeviceIds = Memoizer
            .memoize(WindowsUsbDevice::getControllerDeviceIdList);`
  - In the [`getUsbDevices()` method](https://github.com/oshi/oshi/blob/31858912752ada0e0ceb43b43196ed1409601b96/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java#L137) instead of calling the old method, instead fetch from the memoizer: `List<String> controllerDeviceIdList = controllerDeviceIds.get();`
- Commit your changes, push to to your GitHub fork, and create a Pull Request
- Respond to the usually minor comments on code review
- Celebrate! :tada:

I have a question about a compiling error. I just downloaded the repository onto my device, as well as the most recent versions of jGrasp and Java. When I tried to compile the code, I received many error messages along the lines of "ParseUtil.java:40: error: package org.slf4j does not exist" and other similar errors. Did I clone the project incorrectly? Sorry, I'm really a beginner to Git and open-source projects.

--- a/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java
+++ b/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java
@@ -42,10 +42,9 @@ import com.sun.jna.platform.win32.Cfgmgr32Util;
 import com.sun.jna.platform.win32.COM.WbemcliUtil.WmiQuery;
 import com.sun.jna.platform.win32.COM.WbemcliUtil.WmiResult;
 import com.sun.jna.ptr.IntByReference;
-
+import static oshi.util.Memoizer.memoize;
 import oshi.hardware.UsbDevice;
 import oshi.hardware.common.AbstractUsbDevice;
-import oshi.util.Memoizer;
 import oshi.util.ParseUtil;
 import oshi.util.platform.windows.WmiQueryHandler;
 import oshi.util.platform.windows.WmiUtil;
@@ -57,256 +56,268 @@ import oshi.util.platform.windows.WmiUtil;
  */
 public class WindowsUsbDevice extends AbstractUsbDevice {
 
-	private static final Logger LOG = LoggerFactory.getLogger(WindowsUsbDevice.class);
+    private static final Logger LOG = LoggerFactory.getLogger(WindowsUsbDevice.class);
 
-	enum USBControllerProperty {
-		PNPDEVICEID;
-	}
+    enum USBControllerProperty {
+        PNPDEVICEID;
+    }
 
-	// List of controllers
-	private static Supplier<List<String>> controllerDeviceIds = Memoizer.memoize(WindowsUsbDevice::getControllerDeviceIdList);
+    // List of controllers
+    private static Supplier<List<String>> controllerDeviceIds = memoize(WindowsUsbDevice::getControllerDeviceIdList);
 
-	enum PnPEntityProperty {
-		NAME, MANUFACTURER, PNPDEVICEID;
-	}
+    enum PnPEntityProperty {
+        NAME, MANUFACTURER, PNPDEVICEID;
+    }
 
-	private static final String PNPENTITY_BASE_CLASS = "Win32_PnPEntity";
+    private static final String PNPENTITY_BASE_CLASS = "Win32_PnPEntity";
 
-	enum DiskDriveProperty {
-		PNPDEVICEID, SERIALNUMBER;
-	}
+    enum DiskDriveProperty {
+        PNPDEVICEID, SERIALNUMBER;
+    }
 
-	private static final String DISKDRIVE_BASE_CLASS = "Win32_DiskDrive";
+    private static final String DISKDRIVE_BASE_CLASS = "Win32_DiskDrive";
 
-	private static final Pattern VENDOR_PRODUCT_ID = Pattern
-			.compile(".*(?:VID|VEN)_(\\p{XDigit}{4})&(?:PID|DEV)_(\\p{XDigit}{4}).*");
+    private static final Pattern VENDOR_PRODUCT_ID = Pattern
+            .compile(".*(?:VID|VEN)_(\\p{XDigit}{4})&(?:PID|DEV)_(\\p{XDigit}{4}).*");
 
-	/**
-	 * <p>
-	 * Constructor for WindowsUsbDevice.
-	 * </p>
-	 *
-	 * @param name             a {@link java.lang.String} object.
-	 * @param vendor           a {@link java.lang.String} object.
-	 * @param vendorId         a {@link java.lang.String} object.
-	 * @param productId        a {@link java.lang.String} object.
-	 * @param serialNumber     a {@link java.lang.String} object.
-	 * @param uniqueDeviceId   a {@link java.lang.String} object.
-	 * @param connectedDevices an array of {@link oshi.hardware.UsbDevice} objects.
-	 */
-	public WindowsUsbDevice(String name, String vendor, String vendorId, String productId, String serialNumber,
-			String uniqueDeviceId, UsbDevice[] connectedDevices) {
-		super(name, vendor, vendorId, productId, serialNumber, uniqueDeviceId, connectedDevices);
-	}
+    /**
+     * <p>
+     * Constructor for WindowsUsbDevice.
+     * </p>
+     *
+     * @param name
+     *            a {@link java.lang.String} object.
+     * @param vendor
+     *            a {@link java.lang.String} object.
+     * @param vendorId
+     *            a {@link java.lang.String} object.
+     * @param productId
+     *            a {@link java.lang.String} object.
+     * @param serialNumber
+     *            a {@link java.lang.String} object.
+     * @param uniqueDeviceId
+     *            a {@link java.lang.String} object.
+     * @param connectedDevices
+     *            an array of {@link oshi.hardware.UsbDevice} objects.
+     */
+    public WindowsUsbDevice(String name, String vendor, String vendorId, String productId, String serialNumber,
+            String uniqueDeviceId, UsbDevice[] connectedDevices) {
+        super(name, vendor, vendorId, productId, serialNumber, uniqueDeviceId, connectedDevices);
+    }
 
-	/**
-	 * {@inheritDoc}
-	 *
-	 * @param tree a boolean.
-	 * @return an array of {@link oshi.hardware.UsbDevice} objects.
-	 */
-	public static UsbDevice[] getUsbDevices(boolean tree) {
-		UsbDevice[] devices = getUsbDevices();
-		if (tree) {
-			return devices;
-		}
-		List<UsbDevice> deviceList = new ArrayList<>();
-		// Top level is controllers; they won't be added to the list, but all
-		// their connected devices will be
-		for (UsbDevice device : devices) {
-			addDevicesToList(deviceList, device.getConnectedDevices());
-		}
-		return deviceList.toArray(new UsbDevice[0]);
-	}
+    /**
+     * {@inheritDoc}
+     *
+     * @param tree
+     *            a boolean.
+     * @return an array of {@link oshi.hardware.UsbDevice} objects.
+     */
+    public static UsbDevice[] getUsbDevices(boolean tree) {
+        UsbDevice[] devices = getUsbDevices();
+        if (tree) {
+            return devices;
+        }
+        List<UsbDevice> deviceList = new ArrayList<>();
+        // Top level is controllers; they won't be added to the list, but all
+        // their connected devices will be
+        for (UsbDevice device : devices) {
+            addDevicesToList(deviceList, device.getConnectedDevices());
+        }
+        return deviceList.toArray(new UsbDevice[0]);
+    }
 
-	private static UsbDevice[] getUsbDevices() {
-		// Map to build the recursive tree structure
-		Map<String, List<String>> deviceTreeMap = new HashMap<>();
-		// Track devices seen in the process
-		Set<String> devicesSeen = new HashSet<>();
+    private static UsbDevice[] getUsbDevices() {
+        // Map to build the recursive tree structure
+        Map<String, List<String>> deviceTreeMap = new HashMap<>();
+        // Track devices seen in the process
+        Set<String> devicesSeen = new HashSet<>();
 
-		// Navigate the device tree to track what devices are present
-		List<WindowsUsbDevice> controllerDevices = new ArrayList<>();
-		WmiQueryHandler wmiQueryHandler = WmiQueryHandler.createInstance();
-		List<String> controllerDeviceIdList = controllerDeviceIds.get();
-		for (String controllerDeviceId : controllerDeviceIdList) {
-			putChildrenInDeviceTree(controllerDeviceId, 0, deviceTreeMap, devicesSeen);
-		}
-		// Map to store information using PNPDeviceID as the key.
-		Map<String, WindowsUsbDevice> usbDeviceCache = populateDeviceCache(devicesSeen, wmiQueryHandler);
-		// recursively build results
-		for (String controllerDeviceId : controllerDeviceIdList) {
-			WindowsUsbDevice deviceAndChildren = getDeviceAndChildren(controllerDeviceId, "0000", "0000", deviceTreeMap,
-					usbDeviceCache);
-			if (deviceAndChildren != null) {
-				controllerDevices.add(deviceAndChildren);
-			}
-		}
-		return controllerDevices.toArray(new WindowsUsbDevice[0]);
-	}
+        // Navigate the device tree to track what devices are present
+        List<WindowsUsbDevice> controllerDevices = new ArrayList<>();
+        WmiQueryHandler wmiQueryHandler = WmiQueryHandler.createInstance();
+        List<String> controllerDeviceIdList = controllerDeviceIds.get();
+        for (String controllerDeviceId : controllerDeviceIdList) {
+            putChildrenInDeviceTree(controllerDeviceId, 0, deviceTreeMap, devicesSeen);
+        }
+        // Map to store information using PNPDeviceID as the key.
+        Map<String, WindowsUsbDevice> usbDeviceCache = populateDeviceCache(devicesSeen, wmiQueryHandler);
+        // recursively build results
+        for (String controllerDeviceId : controllerDeviceIdList) {
+            WindowsUsbDevice deviceAndChildren = getDeviceAndChildren(controllerDeviceId, "0000", "0000", deviceTreeMap,
+                    usbDeviceCache);
+            if (deviceAndChildren != null) {
+                controllerDevices.add(deviceAndChildren);
+            }
+        }
+        return controllerDevices.toArray(new WindowsUsbDevice[0]);
+    }
 
-	private static void addDevicesToList(List<UsbDevice> deviceList, UsbDevice[] connectedDevices) {
-		for (UsbDevice device : connectedDevices) {
-			deviceList.add(new WindowsUsbDevice(device.getName(), device.getVendor(), device.getVendorId(),
-					device.getProductId(), device.getSerialNumber(), device.getUniqueDeviceId(), new UsbDevice[0]));
-			addDevicesToList(deviceList, device.getConnectedDevices());
-		}
-	}
+    private static void addDevicesToList(List<UsbDevice> deviceList, UsbDevice[] connectedDevices) {
+        for (UsbDevice device : connectedDevices) {
+            deviceList.add(new WindowsUsbDevice(device.getName(), device.getVendor(), device.getVendorId(),
+                    device.getProductId(), device.getSerialNumber(), device.getUniqueDeviceId(), new UsbDevice[0]));
+            addDevicesToList(deviceList, device.getConnectedDevices());
+        }
+    }
 
-	private static Map<String, WindowsUsbDevice> populateDeviceCache(Set<String> devicesToAdd,
-			WmiQueryHandler wmiQueryHandler) {
-		Map<String, WindowsUsbDevice> usbDeviceCache = new HashMap<>();
-		// Add devices not in the tree
-		if (!devicesToAdd.isEmpty()) {
-			StringBuilder sb = new StringBuilder();
-			boolean first = true;
-			for (String deviceID : devicesToAdd) {
-				if (first) {
-					sb.append(" WHERE (PnPDeviceID=\"");
-					first = false;
-				} else {
-					sb.append(" OR (PnPDeviceID=\"");
-				}
-				sb.append(deviceID).append("\")");
-			}
-			String whereClause = sb.toString();
-			// Query Win32_PnPEntity to populate the maps
-			WmiQuery<PnPEntityProperty> pnpEntityQuery = new WmiQuery<>(PNPENTITY_BASE_CLASS + whereClause,
-					PnPEntityProperty.class);
-			WmiResult<PnPEntityProperty> pnpEntity = wmiQueryHandler.queryWMI(pnpEntityQuery);
-			for (int i = 0; i < pnpEntity.getResultCount(); i++) {
-				String pnpDeviceID = WmiUtil.getString(pnpEntity, PnPEntityProperty.PNPDEVICEID, i);
-				String name = WmiUtil.getString(pnpEntity, PnPEntityProperty.NAME, i);
-				String vendor = WmiUtil.getString(pnpEntity, PnPEntityProperty.MANUFACTURER, i);
-				WindowsUsbDevice device = new WindowsUsbDevice(name, vendor, null, null, "", pnpDeviceID,
-						new WindowsUsbDevice[0]);
-				usbDeviceCache.put(pnpDeviceID, device);
-				LOG.debug("Adding {} to USB device cache.", pnpDeviceID);
-			}
-			// Get serial # for disk drives or other physical media
-			WmiQuery<DiskDriveProperty> diskDriveQuery = new WmiQuery<>(DISKDRIVE_BASE_CLASS + whereClause,
-					DiskDriveProperty.class);
-			WmiResult<DiskDriveProperty> serialNumber = wmiQueryHandler.queryWMI(diskDriveQuery);
-			for (int i = 0; i < serialNumber.getResultCount(); i++) {
-				String pnpDeviceID = WmiUtil.getString(serialNumber, DiskDriveProperty.PNPDEVICEID, i);
-				if (usbDeviceCache.containsKey(pnpDeviceID)) {
-					WindowsUsbDevice device = usbDeviceCache.get(pnpDeviceID);
-					device.serialNumber = ParseUtil
-							.hexStringToString(WmiUtil.getString(serialNumber, DiskDriveProperty.SERIALNUMBER, i));
-				}
-			}
-		}
-		return usbDeviceCache;
-	}
+    private static Map<String, WindowsUsbDevice> populateDeviceCache(Set<String> devicesToAdd,
+            WmiQueryHandler wmiQueryHandler) {
+        Map<String, WindowsUsbDevice> usbDeviceCache = new HashMap<>();
+        // Add devices not in the tree
+        if (!devicesToAdd.isEmpty()) {
+            StringBuilder sb = new StringBuilder();
+            boolean first = true;
+            for (String deviceID : devicesToAdd) {
+                if (first) {
+                    sb.append(" WHERE (PnPDeviceID=\"");
+                    first = false;
+                } else {
+                    sb.append(" OR (PnPDeviceID=\"");
+                }
+                sb.append(deviceID).append("\")");
+            }
+            String whereClause = sb.toString();
+            // Query Win32_PnPEntity to populate the maps
+            WmiQuery<PnPEntityProperty> pnpEntityQuery = new WmiQuery<>(PNPENTITY_BASE_CLASS + whereClause,
+                    PnPEntityProperty.class);
+            WmiResult<PnPEntityProperty> pnpEntity = wmiQueryHandler.queryWMI(pnpEntityQuery);
+            for (int i = 0; i < pnpEntity.getResultCount(); i++) {
+                String pnpDeviceID = WmiUtil.getString(pnpEntity, PnPEntityProperty.PNPDEVICEID, i);
+                String name = WmiUtil.getString(pnpEntity, PnPEntityProperty.NAME, i);
+                String vendor = WmiUtil.getString(pnpEntity, PnPEntityProperty.MANUFACTURER, i);
+                WindowsUsbDevice device = new WindowsUsbDevice(name, vendor, null, null, "", pnpDeviceID,
+                        new WindowsUsbDevice[0]);
+                usbDeviceCache.put(pnpDeviceID, device);
+                LOG.debug("Adding {} to USB device cache.", pnpDeviceID);
+            }
+            // Get serial # for disk drives or other physical media
+            WmiQuery<DiskDriveProperty> diskDriveQuery = new WmiQuery<>(DISKDRIVE_BASE_CLASS + whereClause,
+                    DiskDriveProperty.class);
+            WmiResult<DiskDriveProperty> serialNumber = wmiQueryHandler.queryWMI(diskDriveQuery);
+            for (int i = 0; i < serialNumber.getResultCount(); i++) {
+                String pnpDeviceID = WmiUtil.getString(serialNumber, DiskDriveProperty.PNPDEVICEID, i);
+                if (usbDeviceCache.containsKey(pnpDeviceID)) {
+                    WindowsUsbDevice device = usbDeviceCache.get(pnpDeviceID);
+                    device.serialNumber = ParseUtil
+                            .hexStringToString(WmiUtil.getString(serialNumber, DiskDriveProperty.SERIALNUMBER, i));
+                }
+            }
+        }
+        return usbDeviceCache;
+    }
 
-	/**
-	 * Navigates the Device Tree to place all children PNPDeviceIDs into the map for
-	 * the specified deviceID. Recursively adds children's children, etc.
-	 *
-	 * @param deviceId       The device to add respective children to the map
-	 * @param deviceInstance The device instance (devnode handle), if known. If set
-	 *                       to 0, the code will search for a match.
-	 * @param deviceTreeMap  The overall device tree map that starts at the
-	 *                       controllers
-	 * @param devicesSeen    Devices we've seen so we can add/remove from cache
-	 *                       later
-	 */
-	private static void putChildrenInDeviceTree(String deviceId, int deviceInstance,
-			Map<String, List<String>> deviceTreeMap, Set<String> devicesSeen) {
-		devicesSeen.add(deviceId);
-		// If no devInst provided, find it
-		int devInst = deviceInstance;
-		if (devInst == 0) {
-			IntByReference pdnDevInst = new IntByReference();
-			Cfgmgr32.INSTANCE.CM_Locate_DevNode(pdnDevInst, deviceId, 0);
-			devInst = pdnDevInst.getValue();
-		}
-		// Now iterate the children. Call CM_Get_Child to get first child
-		IntByReference child = new IntByReference();
-		if (0 == Cfgmgr32.INSTANCE.CM_Get_Child(child, devInst, 0)) {
-			// Add first child to a list
-			List<String> childList = new ArrayList<>();
-			String childId = Cfgmgr32Util.CM_Get_Device_ID(child.getValue());
-			childList.add(childId);
-			deviceTreeMap.put(deviceId, childList);
-			putChildrenInDeviceTree(childId, child.getValue(), deviceTreeMap, devicesSeen);
-			// Find any other children
-			IntByReference sibling = new IntByReference();
-			while (0 == Cfgmgr32.INSTANCE.CM_Get_Sibling(sibling, child.getValue(), 0)) {
-				// Add to the list
-				String siblingId = Cfgmgr32Util.CM_Get_Device_ID(sibling.getValue());
-				deviceTreeMap.get(deviceId).add(siblingId);
-				putChildrenInDeviceTree(siblingId, sibling.getValue(), deviceTreeMap, devicesSeen);
-				// Make this sibling the new child to find other siblings
-				child = sibling;
-			}
-		}
-	}
+    /**
+     * Navigates the Device Tree to place all children PNPDeviceIDs into the map for
+     * the specified deviceID. Recursively adds children's children, etc.
+     *
+     * @param deviceId
+     *            The device to add respective children to the map
+     * @param deviceInstance
+     *            The device instance (devnode handle), if known. If set to 0, the
+     *            code will search for a match.
+     * @param deviceTreeMap
+     *            The overall device tree map that starts at the controllers
+     * @param devicesSeen
+     *            Devices we've seen so we can add/remove from cache later
+     */
+    private static void putChildrenInDeviceTree(String deviceId, int deviceInstance,
+            Map<String, List<String>> deviceTreeMap, Set<String> devicesSeen) {
+        devicesSeen.add(deviceId);
+        // If no devInst provided, find it
+        int devInst = deviceInstance;
+        if (devInst == 0) {
+            IntByReference pdnDevInst = new IntByReference();
+            Cfgmgr32.INSTANCE.CM_Locate_DevNode(pdnDevInst, deviceId, 0);
+            devInst = pdnDevInst.getValue();
+        }
+        // Now iterate the children. Call CM_Get_Child to get first child
+        IntByReference child = new IntByReference();
+        if (0 == Cfgmgr32.INSTANCE.CM_Get_Child(child, devInst, 0)) {
+            // Add first child to a list
+            List<String> childList = new ArrayList<>();
+            String childId = Cfgmgr32Util.CM_Get_Device_ID(child.getValue());
+            childList.add(childId);
+            deviceTreeMap.put(deviceId, childList);
+            putChildrenInDeviceTree(childId, child.getValue(), deviceTreeMap, devicesSeen);
+            // Find any other children
+            IntByReference sibling = new IntByReference();
+            while (0 == Cfgmgr32.INSTANCE.CM_Get_Sibling(sibling, child.getValue(), 0)) {
+                // Add to the list
+                String siblingId = Cfgmgr32Util.CM_Get_Device_ID(sibling.getValue());
+                deviceTreeMap.get(deviceId).add(siblingId);
+                putChildrenInDeviceTree(siblingId, sibling.getValue(), deviceTreeMap, devicesSeen);
+                // Make this sibling the new child to find other siblings
+                child = sibling;
+            }
+        }
+    }
 
-	/**
-	 * Recursively creates WindowsUsbDevices by fetching information from maps to
-	 * populate fields
-	 *
-	 * @param hubDeviceId    The PNPdeviceID of this device.
-	 * @param vid            The default (parent) vendor ID
-	 * @param pid            The default (parent) product ID
-	 * @param deviceTreeMap
-	 * @param usbDeviceCache
-	 * @return A WindowsUsbDevice corresponding to this deviceID, or null if unable
-	 *         to find
-	 */
-	private static WindowsUsbDevice getDeviceAndChildren(String hubDeviceId, String vid, String pid,
-			Map<String, List<String>> deviceTreeMap, Map<String, WindowsUsbDevice> usbDeviceCache) {
-		String vendorId = vid;
-		String productId = pid;
-		Matcher m = VENDOR_PRODUCT_ID.matcher(hubDeviceId);
-		if (m.matches()) {
-			vendorId = m.group(1).toLowerCase();
-			productId = m.group(2).toLowerCase();
-		}
-		List<String> pnpDeviceIds = deviceTreeMap.getOrDefault(hubDeviceId, new ArrayList<String>());
-		List<WindowsUsbDevice> usbDevices = new ArrayList<>();
-		for (String pnpDeviceId : pnpDeviceIds) {
-			WindowsUsbDevice deviceAndChildren = getDeviceAndChildren(pnpDeviceId, vendorId, productId, deviceTreeMap,
-					usbDeviceCache);
-			if (deviceAndChildren != null) {
-				usbDevices.add(deviceAndChildren);
-			}
-		}
-		Collections.sort(usbDevices);
-		if (usbDeviceCache.containsKey(hubDeviceId)) {
-			WindowsUsbDevice device = usbDeviceCache.get(hubDeviceId);
-			if (device.name.isEmpty()) {
-				device.name = vendorId + ":" + productId;
-			}
-			device.vendorId = vendorId;
-			device.productId = productId;
-			device.connectedDevices = usbDevices.toArray(new WindowsUsbDevice[0]);
-			return device;
-		}
-		return null;
-	}
+    /**
+     * Recursively creates WindowsUsbDevices by fetching information from maps to
+     * populate fields
+     *
+     * @param hubDeviceId
+     *            The PNPdeviceID of this device.
+     * @param vid
+     *            The default (parent) vendor ID
+     * @param pid
+     *            The default (parent) product ID
+     * @param deviceTreeMap
+     * @param usbDeviceCache
+     * @return A WindowsUsbDevice corresponding to this deviceID, or null if unable
+     *         to find
+     */
+    private static WindowsUsbDevice getDeviceAndChildren(String hubDeviceId, String vid, String pid,
+            Map<String, List<String>> deviceTreeMap, Map<String, WindowsUsbDevice> usbDeviceCache) {
+        String vendorId = vid;
+        String productId = pid;
+        Matcher m = VENDOR_PRODUCT_ID.matcher(hubDeviceId);
+        if (m.matches()) {
+            vendorId = m.group(1).toLowerCase();
+            productId = m.group(2).toLowerCase();
+        }
+        List<String> pnpDeviceIds = deviceTreeMap.getOrDefault(hubDeviceId, new ArrayList<String>());
+        List<WindowsUsbDevice> usbDevices = new ArrayList<>();
+        for (String pnpDeviceId : pnpDeviceIds) {
+            WindowsUsbDevice deviceAndChildren = getDeviceAndChildren(pnpDeviceId, vendorId, productId, deviceTreeMap,
+                    usbDeviceCache);
+            if (deviceAndChildren != null) {
+                usbDevices.add(deviceAndChildren);
+            }
+        }
+        Collections.sort(usbDevices);
+        if (usbDeviceCache.containsKey(hubDeviceId)) {
+            WindowsUsbDevice device = usbDeviceCache.get(hubDeviceId);
+            if (device.name.isEmpty()) {
+                device.name = vendorId + ":" + productId;
+            }
+            device.vendorId = vendorId;
+            device.productId = productId;
+            device.connectedDevices = usbDevices.toArray(new WindowsUsbDevice[0]);
+            return device;
+        }
+        return null;
+    }
 
-	/**
-	 * Queries the USB Controller list, which doesn't change so we only need to
-	 * query it once
-	 *
-	 * @param wmiQueryHandler
-	 *
-	 * @return A list of Strings of USB Controller PNPDeviceIDs
-	 */
-	private static List<String> getControllerDeviceIdList() {
-		WmiQueryHandler wmiQueryHandler = WmiQueryHandler.createInstance();
-		List<String> controllerDeviceIdsList = new ArrayList<>();
-		// One time lookup of USB Controller PnP Device IDs which don't
-		// change
-		WmiQuery<USBControllerProperty> usbControllerQuery = new WmiQuery<>("Win32_USBController",
-				USBControllerProperty.class);
-		WmiResult<USBControllerProperty> usbController = wmiQueryHandler.queryWMI(usbControllerQuery);
-		for (int i = 0; i < usbController.getResultCount(); i++) {
-			controllerDeviceIdsList.add(WmiUtil.getString(usbController, USBControllerProperty.PNPDEVICEID, i));
-		}
+    /**
+     * Queries the USB Controller list, which doesn't change so we only need to
+     * query it once
+     *
+     *
+     * @return A list of Strings of USB Controller PNPDeviceIDs
+     */
+    private static List<String> getControllerDeviceIdList() {
+        WmiQueryHandler wmiQueryHandler = WmiQueryHandler.createInstance();
+        List<String> controllerDeviceIdsList = new ArrayList<>();
+        // One time lookup of USB Controller PnP Device IDs which don't
+        // change
+        WmiQuery<USBControllerProperty> usbControllerQuery = new WmiQuery<>("Win32_USBController",
+                USBControllerProperty.class);
+        WmiResult<USBControllerProperty> usbController = wmiQueryHandler.queryWMI(usbControllerQuery);
+        for (int i = 0; i < usbController.getResultCount(); i++) {
+            controllerDeviceIdsList.add(WmiUtil.getString(usbController, USBControllerProperty.PNPDEVICEID, i));
+        }
 
-		return controllerDeviceIdsList;
-	}
+        return controllerDeviceIdsList;
+    }
 }

--- a/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java
+++ b/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java
@@ -23,6 +23,8 @@
  */
 package oshi.hardware.platform.windows;
 
+import static oshi.util.Memoizer.memoize;
+
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
@@ -42,7 +44,7 @@ import com.sun.jna.platform.win32.Cfgmgr32Util;
 import com.sun.jna.platform.win32.COM.WbemcliUtil.WmiQuery;
 import com.sun.jna.platform.win32.COM.WbemcliUtil.WmiResult;
 import com.sun.jna.ptr.IntByReference;
-import static oshi.util.Memoizer.memoize;
+
 import oshi.hardware.UsbDevice;
 import oshi.hardware.common.AbstractUsbDevice;
 import oshi.util.ParseUtil;
@@ -62,7 +64,6 @@ public class WindowsUsbDevice extends AbstractUsbDevice {
         PNPDEVICEID;
     }
 
-    // List of controllers
     private static Supplier<List<String>> controllerDeviceIds = memoize(WindowsUsbDevice::getControllerDeviceIdList);
 
     enum PnPEntityProperty {
@@ -80,26 +81,6 @@ public class WindowsUsbDevice extends AbstractUsbDevice {
     private static final Pattern VENDOR_PRODUCT_ID = Pattern
             .compile(".*(?:VID|VEN)_(\\p{XDigit}{4})&(?:PID|DEV)_(\\p{XDigit}{4}).*");
 
-    /**
-     * <p>
-     * Constructor for WindowsUsbDevice.
-     * </p>
-     *
-     * @param name
-     *            a {@link java.lang.String} object.
-     * @param vendor
-     *            a {@link java.lang.String} object.
-     * @param vendorId
-     *            a {@link java.lang.String} object.
-     * @param productId
-     *            a {@link java.lang.String} object.
-     * @param serialNumber
-     *            a {@link java.lang.String} object.
-     * @param uniqueDeviceId
-     *            a {@link java.lang.String} object.
-     * @param connectedDevices
-     *            an array of {@link oshi.hardware.UsbDevice} objects.
-     */
     public WindowsUsbDevice(String name, String vendor, String vendorId, String productId, String serialNumber,
             String uniqueDeviceId, UsbDevice[] connectedDevices) {
         super(name, vendor, vendorId, productId, serialNumber, uniqueDeviceId, connectedDevices);

--- a/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java
+++ b/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java
@@ -30,6 +30,7 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.function.Supplier;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
@@ -44,6 +45,7 @@ import com.sun.jna.ptr.IntByReference;
 
 import oshi.hardware.UsbDevice;
 import oshi.hardware.common.AbstractUsbDevice;
+import oshi.util.Memoizer;
 import oshi.util.ParseUtil;
 import oshi.util.platform.windows.WmiQueryHandler;
 import oshi.util.platform.windows.WmiUtil;
@@ -55,269 +57,256 @@ import oshi.util.platform.windows.WmiUtil;
  */
 public class WindowsUsbDevice extends AbstractUsbDevice {
 
-    private static final Logger LOG = LoggerFactory.getLogger(WindowsUsbDevice.class);
+	private static final Logger LOG = LoggerFactory.getLogger(WindowsUsbDevice.class);
 
-    enum USBControllerProperty {
-        PNPDEVICEID;
-    }
+	enum USBControllerProperty {
+		PNPDEVICEID;
+	}
 
-    // List of controllers
-    private static List<String> controllerDeviceIds = null;
+	// List of controllers
+	private static Supplier<List<String>> controllerDeviceIds = Memoizer.memoize(WindowsUsbDevice::getControllerDeviceIdList);
 
-    enum PnPEntityProperty {
-        NAME, MANUFACTURER, PNPDEVICEID;
-    }
+	enum PnPEntityProperty {
+		NAME, MANUFACTURER, PNPDEVICEID;
+	}
 
-    private static final String PNPENTITY_BASE_CLASS = "Win32_PnPEntity";
+	private static final String PNPENTITY_BASE_CLASS = "Win32_PnPEntity";
 
-    enum DiskDriveProperty {
-        PNPDEVICEID, SERIALNUMBER;
-    }
+	enum DiskDriveProperty {
+		PNPDEVICEID, SERIALNUMBER;
+	}
 
-    private static final String DISKDRIVE_BASE_CLASS = "Win32_DiskDrive";
+	private static final String DISKDRIVE_BASE_CLASS = "Win32_DiskDrive";
 
-    private static final Pattern VENDOR_PRODUCT_ID = Pattern
-            .compile(".*(?:VID|VEN)_(\\p{XDigit}{4})&(?:PID|DEV)_(\\p{XDigit}{4}).*");
+	private static final Pattern VENDOR_PRODUCT_ID = Pattern
+			.compile(".*(?:VID|VEN)_(\\p{XDigit}{4})&(?:PID|DEV)_(\\p{XDigit}{4}).*");
 
-    /**
-     * <p>
-     * Constructor for WindowsUsbDevice.
-     * </p>
-     *
-     * @param name
-     *            a {@link java.lang.String} object.
-     * @param vendor
-     *            a {@link java.lang.String} object.
-     * @param vendorId
-     *            a {@link java.lang.String} object.
-     * @param productId
-     *            a {@link java.lang.String} object.
-     * @param serialNumber
-     *            a {@link java.lang.String} object.
-     * @param uniqueDeviceId
-     *            a {@link java.lang.String} object.
-     * @param connectedDevices
-     *            an array of {@link oshi.hardware.UsbDevice} objects.
-     */
-    public WindowsUsbDevice(String name, String vendor, String vendorId, String productId, String serialNumber,
-            String uniqueDeviceId, UsbDevice[] connectedDevices) {
-        super(name, vendor, vendorId, productId, serialNumber, uniqueDeviceId, connectedDevices);
-    }
+	/**
+	 * <p>
+	 * Constructor for WindowsUsbDevice.
+	 * </p>
+	 *
+	 * @param name             a {@link java.lang.String} object.
+	 * @param vendor           a {@link java.lang.String} object.
+	 * @param vendorId         a {@link java.lang.String} object.
+	 * @param productId        a {@link java.lang.String} object.
+	 * @param serialNumber     a {@link java.lang.String} object.
+	 * @param uniqueDeviceId   a {@link java.lang.String} object.
+	 * @param connectedDevices an array of {@link oshi.hardware.UsbDevice} objects.
+	 */
+	public WindowsUsbDevice(String name, String vendor, String vendorId, String productId, String serialNumber,
+			String uniqueDeviceId, UsbDevice[] connectedDevices) {
+		super(name, vendor, vendorId, productId, serialNumber, uniqueDeviceId, connectedDevices);
+	}
 
-    /**
-     * {@inheritDoc}
-     *
-     * @param tree
-     *            a boolean.
-     * @return an array of {@link oshi.hardware.UsbDevice} objects.
-     */
-    public static UsbDevice[] getUsbDevices(boolean tree) {
-        UsbDevice[] devices = getUsbDevices();
-        if (tree) {
-            return devices;
-        }
-        List<UsbDevice> deviceList = new ArrayList<>();
-        // Top level is controllers; they won't be added to the list, but all
-        // their connected devices will be
-        for (UsbDevice device : devices) {
-            addDevicesToList(deviceList, device.getConnectedDevices());
-        }
-        return deviceList.toArray(new UsbDevice[0]);
-    }
+	/**
+	 * {@inheritDoc}
+	 *
+	 * @param tree a boolean.
+	 * @return an array of {@link oshi.hardware.UsbDevice} objects.
+	 */
+	public static UsbDevice[] getUsbDevices(boolean tree) {
+		UsbDevice[] devices = getUsbDevices();
+		if (tree) {
+			return devices;
+		}
+		List<UsbDevice> deviceList = new ArrayList<>();
+		// Top level is controllers; they won't be added to the list, but all
+		// their connected devices will be
+		for (UsbDevice device : devices) {
+			addDevicesToList(deviceList, device.getConnectedDevices());
+		}
+		return deviceList.toArray(new UsbDevice[0]);
+	}
 
-    private static UsbDevice[] getUsbDevices() {
-        // Map to build the recursive tree structure
-        Map<String, List<String>> deviceTreeMap = new HashMap<>();
-        // Track devices seen in the process
-        Set<String> devicesSeen = new HashSet<>();
+	private static UsbDevice[] getUsbDevices() {
+		// Map to build the recursive tree structure
+		Map<String, List<String>> deviceTreeMap = new HashMap<>();
+		// Track devices seen in the process
+		Set<String> devicesSeen = new HashSet<>();
 
-        // Navigate the device tree to track what devices are present
-        List<WindowsUsbDevice> controllerDevices = new ArrayList<>();
-        WmiQueryHandler wmiQueryHandler = WmiQueryHandler.createInstance();
-        List<String> controllerDeviceIdList = getControllerDeviceIdList(wmiQueryHandler);
-        for (String controllerDeviceId : controllerDeviceIdList) {
-            putChildrenInDeviceTree(controllerDeviceId, 0, deviceTreeMap, devicesSeen);
-        }
-        // Map to store information using PNPDeviceID as the key.
-        Map<String, WindowsUsbDevice> usbDeviceCache = populateDeviceCache(devicesSeen, wmiQueryHandler);
-        // recursively build results
-        for (String controllerDeviceId : controllerDeviceIdList) {
-            WindowsUsbDevice deviceAndChildren = getDeviceAndChildren(controllerDeviceId, "0000", "0000", deviceTreeMap,
-                    usbDeviceCache);
-            if (deviceAndChildren != null) {
-                controllerDevices.add(deviceAndChildren);
-            }
-        }
-        return controllerDevices.toArray(new WindowsUsbDevice[0]);
-    }
+		// Navigate the device tree to track what devices are present
+		List<WindowsUsbDevice> controllerDevices = new ArrayList<>();
+		WmiQueryHandler wmiQueryHandler = WmiQueryHandler.createInstance();
+		List<String> controllerDeviceIdList = controllerDeviceIds.get();
+		for (String controllerDeviceId : controllerDeviceIdList) {
+			putChildrenInDeviceTree(controllerDeviceId, 0, deviceTreeMap, devicesSeen);
+		}
+		// Map to store information using PNPDeviceID as the key.
+		Map<String, WindowsUsbDevice> usbDeviceCache = populateDeviceCache(devicesSeen, wmiQueryHandler);
+		// recursively build results
+		for (String controllerDeviceId : controllerDeviceIdList) {
+			WindowsUsbDevice deviceAndChildren = getDeviceAndChildren(controllerDeviceId, "0000", "0000", deviceTreeMap,
+					usbDeviceCache);
+			if (deviceAndChildren != null) {
+				controllerDevices.add(deviceAndChildren);
+			}
+		}
+		return controllerDevices.toArray(new WindowsUsbDevice[0]);
+	}
 
-    private static void addDevicesToList(List<UsbDevice> deviceList, UsbDevice[] connectedDevices) {
-        for (UsbDevice device : connectedDevices) {
-            deviceList.add(new WindowsUsbDevice(device.getName(), device.getVendor(), device.getVendorId(),
-                    device.getProductId(), device.getSerialNumber(), device.getUniqueDeviceId(), new UsbDevice[0]));
-            addDevicesToList(deviceList, device.getConnectedDevices());
-        }
-    }
+	private static void addDevicesToList(List<UsbDevice> deviceList, UsbDevice[] connectedDevices) {
+		for (UsbDevice device : connectedDevices) {
+			deviceList.add(new WindowsUsbDevice(device.getName(), device.getVendor(), device.getVendorId(),
+					device.getProductId(), device.getSerialNumber(), device.getUniqueDeviceId(), new UsbDevice[0]));
+			addDevicesToList(deviceList, device.getConnectedDevices());
+		}
+	}
 
-    private static Map<String, WindowsUsbDevice> populateDeviceCache(Set<String> devicesToAdd,
-            WmiQueryHandler wmiQueryHandler) {
-        Map<String, WindowsUsbDevice> usbDeviceCache = new HashMap<>();
-        // Add devices not in the tree
-        if (!devicesToAdd.isEmpty()) {
-            StringBuilder sb = new StringBuilder();
-            boolean first = true;
-            for (String deviceID : devicesToAdd) {
-                if (first) {
-                    sb.append(" WHERE (PnPDeviceID=\"");
-                    first = false;
-                } else {
-                    sb.append(" OR (PnPDeviceID=\"");
-                }
-                sb.append(deviceID).append("\")");
-            }
-            String whereClause = sb.toString();
-            // Query Win32_PnPEntity to populate the maps
-            WmiQuery<PnPEntityProperty> pnpEntityQuery = new WmiQuery<>(PNPENTITY_BASE_CLASS + whereClause,
-                    PnPEntityProperty.class);
-            WmiResult<PnPEntityProperty> pnpEntity = wmiQueryHandler.queryWMI(pnpEntityQuery);
-            for (int i = 0; i < pnpEntity.getResultCount(); i++) {
-                String pnpDeviceID = WmiUtil.getString(pnpEntity, PnPEntityProperty.PNPDEVICEID, i);
-                String name = WmiUtil.getString(pnpEntity, PnPEntityProperty.NAME, i);
-                String vendor = WmiUtil.getString(pnpEntity, PnPEntityProperty.MANUFACTURER, i);
-                WindowsUsbDevice device = new WindowsUsbDevice(name, vendor, null, null, "", pnpDeviceID,
-                        new WindowsUsbDevice[0]);
-                usbDeviceCache.put(pnpDeviceID, device);
-                LOG.debug("Adding {} to USB device cache.", pnpDeviceID);
-            }
-            // Get serial # for disk drives or other physical media
-            WmiQuery<DiskDriveProperty> diskDriveQuery = new WmiQuery<>(DISKDRIVE_BASE_CLASS + whereClause,
-                    DiskDriveProperty.class);
-            WmiResult<DiskDriveProperty> serialNumber = wmiQueryHandler.queryWMI(diskDriveQuery);
-            for (int i = 0; i < serialNumber.getResultCount(); i++) {
-                String pnpDeviceID = WmiUtil.getString(serialNumber, DiskDriveProperty.PNPDEVICEID, i);
-                if (usbDeviceCache.containsKey(pnpDeviceID)) {
-                    WindowsUsbDevice device = usbDeviceCache.get(pnpDeviceID);
-                    device.serialNumber = ParseUtil
-                            .hexStringToString(WmiUtil.getString(serialNumber, DiskDriveProperty.SERIALNUMBER, i));
-                }
-            }
-        }
-        return usbDeviceCache;
-    }
+	private static Map<String, WindowsUsbDevice> populateDeviceCache(Set<String> devicesToAdd,
+			WmiQueryHandler wmiQueryHandler) {
+		Map<String, WindowsUsbDevice> usbDeviceCache = new HashMap<>();
+		// Add devices not in the tree
+		if (!devicesToAdd.isEmpty()) {
+			StringBuilder sb = new StringBuilder();
+			boolean first = true;
+			for (String deviceID : devicesToAdd) {
+				if (first) {
+					sb.append(" WHERE (PnPDeviceID=\"");
+					first = false;
+				} else {
+					sb.append(" OR (PnPDeviceID=\"");
+				}
+				sb.append(deviceID).append("\")");
+			}
+			String whereClause = sb.toString();
+			// Query Win32_PnPEntity to populate the maps
+			WmiQuery<PnPEntityProperty> pnpEntityQuery = new WmiQuery<>(PNPENTITY_BASE_CLASS + whereClause,
+					PnPEntityProperty.class);
+			WmiResult<PnPEntityProperty> pnpEntity = wmiQueryHandler.queryWMI(pnpEntityQuery);
+			for (int i = 0; i < pnpEntity.getResultCount(); i++) {
+				String pnpDeviceID = WmiUtil.getString(pnpEntity, PnPEntityProperty.PNPDEVICEID, i);
+				String name = WmiUtil.getString(pnpEntity, PnPEntityProperty.NAME, i);
+				String vendor = WmiUtil.getString(pnpEntity, PnPEntityProperty.MANUFACTURER, i);
+				WindowsUsbDevice device = new WindowsUsbDevice(name, vendor, null, null, "", pnpDeviceID,
+						new WindowsUsbDevice[0]);
+				usbDeviceCache.put(pnpDeviceID, device);
+				LOG.debug("Adding {} to USB device cache.", pnpDeviceID);
+			}
+			// Get serial # for disk drives or other physical media
+			WmiQuery<DiskDriveProperty> diskDriveQuery = new WmiQuery<>(DISKDRIVE_BASE_CLASS + whereClause,
+					DiskDriveProperty.class);
+			WmiResult<DiskDriveProperty> serialNumber = wmiQueryHandler.queryWMI(diskDriveQuery);
+			for (int i = 0; i < serialNumber.getResultCount(); i++) {
+				String pnpDeviceID = WmiUtil.getString(serialNumber, DiskDriveProperty.PNPDEVICEID, i);
+				if (usbDeviceCache.containsKey(pnpDeviceID)) {
+					WindowsUsbDevice device = usbDeviceCache.get(pnpDeviceID);
+					device.serialNumber = ParseUtil
+							.hexStringToString(WmiUtil.getString(serialNumber, DiskDriveProperty.SERIALNUMBER, i));
+				}
+			}
+		}
+		return usbDeviceCache;
+	}
 
-    /**
-     * Navigates the Device Tree to place all children PNPDeviceIDs into the map for
-     * the specified deviceID. Recursively adds children's children, etc.
-     *
-     * @param deviceId
-     *            The device to add respective children to the map
-     * @param deviceInstance
-     *            The device instance (devnode handle), if known. If set to 0, the
-     *            code will search for a match.
-     * @param deviceTreeMap
-     *            The overall device tree map that starts at the controllers
-     * @param devicesSeen
-     *            Devices we've seen so we can add/remove from cache later
-     */
-    private static void putChildrenInDeviceTree(String deviceId, int deviceInstance,
-            Map<String, List<String>> deviceTreeMap, Set<String> devicesSeen) {
-        devicesSeen.add(deviceId);
-        // If no devInst provided, find it
-        int devInst = deviceInstance;
-        if (devInst == 0) {
-            IntByReference pdnDevInst = new IntByReference();
-            Cfgmgr32.INSTANCE.CM_Locate_DevNode(pdnDevInst, deviceId, 0);
-            devInst = pdnDevInst.getValue();
-        }
-        // Now iterate the children. Call CM_Get_Child to get first child
-        IntByReference child = new IntByReference();
-        if (0 == Cfgmgr32.INSTANCE.CM_Get_Child(child, devInst, 0)) {
-            // Add first child to a list
-            List<String> childList = new ArrayList<>();
-            String childId = Cfgmgr32Util.CM_Get_Device_ID(child.getValue());
-            childList.add(childId);
-            deviceTreeMap.put(deviceId, childList);
-            putChildrenInDeviceTree(childId, child.getValue(), deviceTreeMap, devicesSeen);
-            // Find any other children
-            IntByReference sibling = new IntByReference();
-            while (0 == Cfgmgr32.INSTANCE.CM_Get_Sibling(sibling, child.getValue(), 0)) {
-                // Add to the list
-                String siblingId = Cfgmgr32Util.CM_Get_Device_ID(sibling.getValue());
-                deviceTreeMap.get(deviceId).add(siblingId);
-                putChildrenInDeviceTree(siblingId, sibling.getValue(), deviceTreeMap, devicesSeen);
-                // Make this sibling the new child to find other siblings
-                child = sibling;
-            }
-        }
-    }
+	/**
+	 * Navigates the Device Tree to place all children PNPDeviceIDs into the map for
+	 * the specified deviceID. Recursively adds children's children, etc.
+	 *
+	 * @param deviceId       The device to add respective children to the map
+	 * @param deviceInstance The device instance (devnode handle), if known. If set
+	 *                       to 0, the code will search for a match.
+	 * @param deviceTreeMap  The overall device tree map that starts at the
+	 *                       controllers
+	 * @param devicesSeen    Devices we've seen so we can add/remove from cache
+	 *                       later
+	 */
+	private static void putChildrenInDeviceTree(String deviceId, int deviceInstance,
+			Map<String, List<String>> deviceTreeMap, Set<String> devicesSeen) {
+		devicesSeen.add(deviceId);
+		// If no devInst provided, find it
+		int devInst = deviceInstance;
+		if (devInst == 0) {
+			IntByReference pdnDevInst = new IntByReference();
+			Cfgmgr32.INSTANCE.CM_Locate_DevNode(pdnDevInst, deviceId, 0);
+			devInst = pdnDevInst.getValue();
+		}
+		// Now iterate the children. Call CM_Get_Child to get first child
+		IntByReference child = new IntByReference();
+		if (0 == Cfgmgr32.INSTANCE.CM_Get_Child(child, devInst, 0)) {
+			// Add first child to a list
+			List<String> childList = new ArrayList<>();
+			String childId = Cfgmgr32Util.CM_Get_Device_ID(child.getValue());
+			childList.add(childId);
+			deviceTreeMap.put(deviceId, childList);
+			putChildrenInDeviceTree(childId, child.getValue(), deviceTreeMap, devicesSeen);
+			// Find any other children
+			IntByReference sibling = new IntByReference();
+			while (0 == Cfgmgr32.INSTANCE.CM_Get_Sibling(sibling, child.getValue(), 0)) {
+				// Add to the list
+				String siblingId = Cfgmgr32Util.CM_Get_Device_ID(sibling.getValue());
+				deviceTreeMap.get(deviceId).add(siblingId);
+				putChildrenInDeviceTree(siblingId, sibling.getValue(), deviceTreeMap, devicesSeen);
+				// Make this sibling the new child to find other siblings
+				child = sibling;
+			}
+		}
+	}
 
-    /**
-     * Recursively creates WindowsUsbDevices by fetching information from maps to
-     * populate fields
-     *
-     * @param hubDeviceId
-     *            The PNPdeviceID of this device.
-     * @param vid
-     *            The default (parent) vendor ID
-     * @param pid
-     *            The default (parent) product ID
-     * @param deviceTreeMap
-     * @param usbDeviceCache
-     * @return A WindowsUsbDevice corresponding to this deviceID, or null if unable
-     *         to find
-     */
-    private static WindowsUsbDevice getDeviceAndChildren(String hubDeviceId, String vid, String pid,
-            Map<String, List<String>> deviceTreeMap, Map<String, WindowsUsbDevice> usbDeviceCache) {
-        String vendorId = vid;
-        String productId = pid;
-        Matcher m = VENDOR_PRODUCT_ID.matcher(hubDeviceId);
-        if (m.matches()) {
-            vendorId = m.group(1).toLowerCase();
-            productId = m.group(2).toLowerCase();
-        }
-        List<String> pnpDeviceIds = deviceTreeMap.getOrDefault(hubDeviceId, new ArrayList<String>());
-        List<WindowsUsbDevice> usbDevices = new ArrayList<>();
-        for (String pnpDeviceId : pnpDeviceIds) {
-            WindowsUsbDevice deviceAndChildren = getDeviceAndChildren(pnpDeviceId, vendorId, productId, deviceTreeMap,
-                    usbDeviceCache);
-            if (deviceAndChildren != null) {
-                usbDevices.add(deviceAndChildren);
-            }
-        }
-        Collections.sort(usbDevices);
-        if (usbDeviceCache.containsKey(hubDeviceId)) {
-            WindowsUsbDevice device = usbDeviceCache.get(hubDeviceId);
-            if (device.name.isEmpty()) {
-                device.name = vendorId + ":" + productId;
-            }
-            device.vendorId = vendorId;
-            device.productId = productId;
-            device.connectedDevices = usbDevices.toArray(new WindowsUsbDevice[0]);
-            return device;
-        }
-        return null;
-    }
+	/**
+	 * Recursively creates WindowsUsbDevices by fetching information from maps to
+	 * populate fields
+	 *
+	 * @param hubDeviceId    The PNPdeviceID of this device.
+	 * @param vid            The default (parent) vendor ID
+	 * @param pid            The default (parent) product ID
+	 * @param deviceTreeMap
+	 * @param usbDeviceCache
+	 * @return A WindowsUsbDevice corresponding to this deviceID, or null if unable
+	 *         to find
+	 */
+	private static WindowsUsbDevice getDeviceAndChildren(String hubDeviceId, String vid, String pid,
+			Map<String, List<String>> deviceTreeMap, Map<String, WindowsUsbDevice> usbDeviceCache) {
+		String vendorId = vid;
+		String productId = pid;
+		Matcher m = VENDOR_PRODUCT_ID.matcher(hubDeviceId);
+		if (m.matches()) {
+			vendorId = m.group(1).toLowerCase();
+			productId = m.group(2).toLowerCase();
+		}
+		List<String> pnpDeviceIds = deviceTreeMap.getOrDefault(hubDeviceId, new ArrayList<String>());
+		List<WindowsUsbDevice> usbDevices = new ArrayList<>();
+		for (String pnpDeviceId : pnpDeviceIds) {
+			WindowsUsbDevice deviceAndChildren = getDeviceAndChildren(pnpDeviceId, vendorId, productId, deviceTreeMap,
+					usbDeviceCache);
+			if (deviceAndChildren != null) {
+				usbDevices.add(deviceAndChildren);
+			}
+		}
+		Collections.sort(usbDevices);
+		if (usbDeviceCache.containsKey(hubDeviceId)) {
+			WindowsUsbDevice device = usbDeviceCache.get(hubDeviceId);
+			if (device.name.isEmpty()) {
+				device.name = vendorId + ":" + productId;
+			}
+			device.vendorId = vendorId;
+			device.productId = productId;
+			device.connectedDevices = usbDevices.toArray(new WindowsUsbDevice[0]);
+			return device;
+		}
+		return null;
+	}
 
-    /**
-     * Queries the USB Controller list, which doesn't change so we only need to
-     * query it once
-     *
-     * @param wmiQueryHandler
-     *
-     * @return A list of Strings of USB Controller PNPDeviceIDs
-     */
-    private static List<String> getControllerDeviceIdList(WmiQueryHandler wmiQueryHandler) {
-        if (controllerDeviceIds == null) {
-            controllerDeviceIds = new ArrayList<>();
-            // One time lookup of USB Controller PnP Device IDs which don't
-            // change
-            WmiQuery<USBControllerProperty> usbControllerQuery = new WmiQuery<>("Win32_USBController",
-                    USBControllerProperty.class);
-            WmiResult<USBControllerProperty> usbController = wmiQueryHandler.queryWMI(usbControllerQuery);
-            for (int i = 0; i < usbController.getResultCount(); i++) {
-                controllerDeviceIds.add(WmiUtil.getString(usbController, USBControllerProperty.PNPDEVICEID, i));
-            }
-        }
-        return controllerDeviceIds;
-    }
+	/**
+	 * Queries the USB Controller list, which doesn't change so we only need to
+	 * query it once
+	 *
+	 * @param wmiQueryHandler
+	 *
+	 * @return A list of Strings of USB Controller PNPDeviceIDs
+	 */
+	private static List<String> getControllerDeviceIdList() {
+		WmiQueryHandler wmiQueryHandler = WmiQueryHandler.createInstance();
+		List<String> controllerDeviceIdsList = new ArrayList<>();
+		// One time lookup of USB Controller PnP Device IDs which don't
+		// change
+		WmiQuery<USBControllerProperty> usbControllerQuery = new WmiQuery<>("Win32_USBController",
+				USBControllerProperty.class);
+		WmiResult<USBControllerProperty> usbController = wmiQueryHandler.queryWMI(usbControllerQuery);
+		for (int i = 0; i < usbController.getResultCount(); i++) {
+			controllerDeviceIdsList.add(WmiUtil.getString(usbController, USBControllerProperty.PNPDEVICEID, i));
+		}
+
+		return controllerDeviceIdsList;
+	}
 }

